### PR TITLE
[ONNX] Quantization operators

### PR DESCRIPTION
**Pull request by tsocha**
Add support for quantized ONNX ops, expected in the next ONNX revision.

### CODE REVIEW

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: Please update the year.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.hpp]: Please update the year.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.hpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace ngraph
+{
+    namespace onnx_import
+    {
+        namespace op
+        {
+            namespace set_1
+            {
+                ngraph::NodeVector quantize_linear(const ngraph::onnx_import::Node& node);
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.hpp]: We're already in the right namespace, there's no need to use fully qualified name.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: Please update the year.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace ngraph
+{
+    namespace onnx_import
+    {
+        namespace op
+        {
+            namespace set_1
+            {
+                ngraph::NodeVector quantize_linear(const ngraph::onnx_import::Node& node)
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: We're already in the right namespace, there's no need to use fully qualified name.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                {
+                    ngraph::NodeVector inputs{node.get_ng_inputs()};
+                    std::shared_ptr<ngraph::Node> x = inputs.at(0);
+                    std::shared_ptr<ngraph::Node> y_scale = inputs.at(1);
+                    std::shared_ptr<ngraph::Node> y_zero_point = inputs.at(2);
+
+                    bool has_axis = false;
+                    ngraph::Shape x_shape = x->get_shape();
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: Similarly here, no need to use `ngraph::` namespace. `Shape` and `AxisSet` object are unique in `ngraph` namespace we're already in.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                    ngraph::Shape y_zero_point_shape = y_zero_point->get_shape();
+                    ngraph::AxisSet axis_set{};
+
+                    try
+                    {
+                        std::int64_t axis = node.get_attribute_value<std::int64_t>("axis");
+                        axis_set = ngraph::AxisSet{static_cast<std::size_t>(axis)};
+                        has_axis = true;
+                    }
+                    catch (const ngraph::onnx_import::error::node::UnknownAttribute&)
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: Namespace

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                            << "y_zero_point must be a scalar if no axis is provided.";
+                    }
+
+                    return {std::make_shared<ngraph::op::Quantize>(
+                        x,
+                        y_scale,
+                        y_zero_point,
+                        y_zero_point->get_element_type(),
+                        axis_set,
+                        // TODO (arogowiec) : Get rounding mode from ONNX instead of hard-coding
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: What about this comment? AFAIK there hasn't been any ONNX attribute related to this. Please check on Microsoft's onnxruntime github in contrib_ops directory.

File: src/ngraph/frontend/onnx_import/op/quant_conv.hpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.hpp]: Please update the year.

File: src/ngraph/frontend/onnx_import/op/quant_conv.hpp [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            namespace set_1
+            {
+                /// \brief Performs ONNX Quant Conv operation.
+                ///
+                /// \param node   The ONNX node object representing this operation.
+                ///
+                /// \return The vector containing Ngraph nodes producing output of ONNX quantizied convolution
+                ///         operation.
+                ngraph::NodeVector quant_conv(const ngraph::onnx_import::Node& node);
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.hpp]: We're already in the right namespace, there's no need to use fully qualified name.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 11]
```diff
@@ -0,0 +1,248 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Please update the year.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+                        std::shared_ptr<ngraph::Node> data_zero_point;
+                        std::shared_ptr<ngraph::Node> filter_zero_point;
+                        std::shared_ptr<ngraph::Node> output_zero_point;
+                    };
+
+                    std::shared_ptr<ngraph::Node>
+                        make_ng_quant_conv(const std::shared_ptr<ngraph::Node>& data,
+                                           const std::shared_ptr<ngraph::Node>& filters,
+                                           const ngraph::Strides& strides,
+                                           const ngraph::Strides& filter_dilations,
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Add header.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+                                           const ngraph::CoordinateDiff& padding_below,
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Add header.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+                                           const OpZeroPoint& op_zero_point)
+                    {
+                        if (groups > 1)
+                        {
+                            // Split one convolution op to N ops where N is the number of groups
+                            // and concat results after computation.
+                            // reference: https://github.com/NervanaSystems/ngraph-mxnet/blob/fdd692/src/ngraph/ngraph_emitter.cc#L822-L856
+                            std::size_t n_data_channels{data->get_shape().at(1)};
+                            std::size_t n_filters_channels{filters->get_shape().at(0)};
+                            // TODO: ensure n_data_channels % groups = 0
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Add check from this comment. This should have been already added in conventional convolutions.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+                    ASSERT_VALID_ARGUMENT(node,
+                                          ((groups >= 0) && (groups <= data->get_shape().at(1)) &&
+                                           (groups <= filters->get_shape().at(0))))
+                        << "incorrect value of 'group' attribute: " << groups;
+
+                    ngraph::Strides strides = ngraph::onnx_import::convpool::get_strides(node);
+                    ngraph::Strides filter_dilations =
+                        ngraph::onnx_import::convpool::get_dilations(node);
+                    ngraph::Strides data_dilations = ngraph::Strides(
+                        ngraph::onnx_import::convpool::get_kernel_shape(node).size(), 1UL);
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: No need for fully qualified namespace. Same above.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+                                                const ngraph::CoordinateDiff& padding_above,
+                                                const ngraph::Strides& data_dilations,
+                                                int groups,
+                                                const OpScale& op_scale,
+                                                const OpZeroPoint& op_zero_point)
+                    {
+                        if (groups > 1)
+                        {
+                            throw ngraph::onnx_import::error::NotSupported(
+                                "No support for quantized group conv+bias.");
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Please check whether this is still unsupported?

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.hpp [hunk 17]
```diff
@@ -0,0 +1,38 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.hpp]: Please update the year.

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.hpp [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace ngraph
+{
+    namespace onnx_import
+    {
+        namespace op
+        {
+            namespace set_1
+            {
+                ngraph::NodeVector dequantize_linear(const ngraph::onnx_import::Node& node);
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.hpp]: Namespace.

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+namespace ngraph
+{
+    namespace onnx_import
+    {
+        namespace op
+        {
+            namespace set_1
+            {
+                ngraph::NodeVector dequantize_linear(const ngraph::onnx_import::Node& node)
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: No need for fully qualified namespace.

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+                    ngraph::NodeVector inputs{node.get_ng_inputs()};
+                    std::shared_ptr<ngraph::Node> x = inputs.at(0);
+                    std::shared_ptr<ngraph::Node> y_scale = inputs.at(1);
+                    std::shared_ptr<ngraph::Node> y_zero_point = inputs.at(2);
+
+                    bool has_axis = false;
+                    ngraph::Shape x_shape = x->get_shape();
+                    ngraph::Shape y_scale_shape = y_scale->get_shape();
+                    ngraph::Shape y_zero_point_shape = y_zero_point->get_shape();
+                    ngraph::AxisSet axis_set{};
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: Namespace.

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+                    // Per tensor quantization. Input ‘scale’ and ‘zero_point’ must be scalars
+                    else
+                    {
+                        ASSERT_VALID_ARGUMENT(node, y_scale_shape.size() == 0)
+                            << "y_scale must be a scalar if no axis is provided.";
+                        ASSERT_VALID_ARGUMENT(node, y_zero_point_shape.size() == 0)
+                            << "y_zero_point must be a scalar if no axis is provided.";
+                    }
+
+                    // TODO:  THIS IS A WORKAROUND WHICH SHOULD BE REMOVED
```
Comment (arogowie-intel) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: Inspect whether this is still needed.

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp (hunk 8 above)
Comment (tsocha) [reply src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: @adstraw Should we remove this `TODO` if documentation say nothing about rounding mode?

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp (hunk 16 above)
Comment (tsocha) [reply src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: @adstraw should we try to fix that or it will be supported from different place?

**Comment by adstraw** [comment]
Adding Ashok to this review.  Ashok will make sure this commit runs our latest models with good perf and accuracy.

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+                        if (groups > 1)
+                        {
+                            // Split one convolution op to N ops where N is the number of groups
+                            // and concat results after computation.
+                            // reference: https://github.com/NervanaSystems/ngraph-mxnet/blob/fdd692/src/ngraph/ngraph_emitter.cc#L822-L856
+                            std::size_t n_data_channels{data->get_shape().at(1)};
+                            std::size_t n_filters_channels{filters->get_shape().at(0)};
+
+                            std::size_t data_group_size{n_data_channels / groups};
+                            std::size_t filters_group_size{n_filters_channels / groups};
```
Comment (adstraw) [inline src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Is the number of data/filters modulo groups guaranteed to be zero?

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+                    {
+                        std::int64_t axis = node.get_attribute_value<std::int64_t>("axis");
+                        axis_set = AxisSet{static_cast<std::size_t>(axis)};
+                        has_axis = true;
+                    }
+                    catch (const error::node::UnknownAttribute&)
+                    {
+                    }
+
+                    // Per `axis` quantization. Input ‘scale’ and ‘zero_point’ must be 1-D tensors
```
Comment (adstraw) [inline src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: quite a lot of duplicated code between quantize and dequantize.  could you reuse some of this code in a function?

File: src/ngraph/frontend/onnx_import/op/quantize_linear.cpp (hunk 8 above)
Comment (adstraw) [reply src/ngraph/frontend/onnx_import/op/quantize_linear.cpp]: Latest spec (which I also put on Slack) says:

"The quantization formula is y = (x / y_scale) + y_zero_point. For (x / y_scale), it’s rounding to the nearest integer value (in floating-point format) and rounding half to even."

So, I think you can remove this TODO.

**Comment by postrational** [comment]
@ashokei Please not this requires models which assume quant ops are in the `ai.onnx` domain. You can simply remove the `domain` field from all nodes to get this.
```python
import onnx
model = onnx.load_model('/path/to/model.onnx')
for node in model.graph.node:
    node.ClearField('domain')
onnx.save_model(model, '/path/to/modified_model.onnx')
```

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp (hunk 22 above)
Comment (tsocha) [reply src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: Yes, int the line 200 we have assert.

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+                    // Per tensor quantization. Input ‘scale’ and ‘zero_point’ must be scalars
+                    else
+                    {
+                        ASSERT_VALID_ARGUMENT(node, y_scale_shape.size() == 0)
+                            << "y_scale must be a scalar if no axis is provided.";
+                        ASSERT_VALID_ARGUMENT(node, y_zero_point_shape.size() == 0)
+                            << "y_zero_point must be a scalar if no axis is provided.";
+                    }
+
+                    return {std::make_shared<ngraph::op::Dequantize>(
```
Comment (ashokei) [inline src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: dequantize in CPU/IAT backend expects constant zero, we need below check,
```
if (x->get_element_type() != y_zero_point->get_element_type())                        
{                                                                                     
    // Currently only support Constant node.                                          
    ASSERT_IS_SUPPORTED(node, y_zero_point->is_constant())                            
        << "doesn't support zero point input of other type than Constant.";           
                                                                                      
    auto y_zp_constant = std::static_pointer_cast<ngraph::op::Constant>(y_zero_point);
                                                                                      
    y_zero_point = ngraph::op::Constant::create<std::uint8_t>(                        
        x->get_element_type(),                                                        
        y_zero_point->get_shape(),                                                    
        y_zp_constant->get_vector<std::uint8_t>());
}
```

File: src/ngraph/frontend/onnx_import/op/quant_conv.cpp (hunk 22 above)
Comment (adstraw) [reply src/ngraph/frontend/onnx_import/op/quant_conv.cpp]: missed that; thanks.

File: src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp (hunk 24 above)
Comment (ashokei) [reply src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: after discussion with @adstraw , it seems this check can be done on CPU/IAT backend side.

src/ngraph/runtime/cpu/pass/cpu_assignment.cpp:754

Comment (ashokei) [reply src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: dequantize core API expects `
x->get_element_type() == y_zero_point->get_element_type()` , If this is not the case, we will need to recreate zero with same type as in above snippet. @adstraw maybe we need above check in this PR ?

Comment (tsocha) [reply src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: Is converting `offset` to `data` element type in this CPU pass is a good idea?
Or should we convert it in the onnx bridge?
If `offset == 0` is checked in pass already.

Comment (ashokei) [reply src/ngraph/frontend/onnx_import/op/dequantize_linear.cpp]: `convert it in the onnx bridge` is better, as this is API constraint/requirement.

**Pull request merged**